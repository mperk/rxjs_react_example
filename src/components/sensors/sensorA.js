import { map, timer } from 'rxjs'
import { getRandomInt } from '../../utils'
const name = "Sensor A"
const sensorA = {
    create: () => {
        let intervalDuration = getRandomInt(200, 1500)
        // console.log("A :" + intervalDuration)
        return timer(0, intervalDuration)
            .pipe(
                map(i => ({
                    name: name,
                    data: Math.random().toString(36).slice(-8),
                    lastRunTime: new Date() - intervalDuration
                }))
            )
    }
}

export default sensorA