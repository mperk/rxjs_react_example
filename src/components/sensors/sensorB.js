import { map, timer } from 'rxjs'
import { getRandomInt } from '../../utils'
const name = "Sensor B"
const sensorB = {
    create: () => {
        let intervalDuration = getRandomInt(200, 1500)
        return timer(0, intervalDuration)
            .pipe(
                map(i => ({
                    name: name,
                    data: Math.random().toString(36).slice(-8),
                    lastRunTime: new Date() - intervalDuration
                }))
            )
    }
}

export default sensorB
