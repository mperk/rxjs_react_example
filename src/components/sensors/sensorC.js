import { map, timer } from 'rxjs'
import { getRandomInt } from '../../utils'
const name = "Sensor C"
const sensorC = {
    create: () => {
        let intervalDuration = getRandomInt(200, 1500)
        return timer(0, intervalDuration)
            .pipe(
                map(i => ({
                    name: name,
                    data: Math.random().toString(36).slice(-8),
                    lastRunTime: new Date() - intervalDuration
                }))
            )
    }
}

export default sensorC
