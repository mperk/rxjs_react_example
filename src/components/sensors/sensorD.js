import { map, timer } from 'rxjs'
import { getRandomInt } from '../../utils'
const name = "Sensor D"
const sensorD = {
    create: () => {
        let intervalDuration = getRandomInt(200, 1500)
        // let intervalDuration = 1400
        return timer(0, intervalDuration)
            .pipe(
                map(i => ({
                    name: name,
                    data: Math.random().toString(36).slice(-8),
                    lastRunTime: new Date() - intervalDuration
                }))
            )
    }
}

export default sensorD
