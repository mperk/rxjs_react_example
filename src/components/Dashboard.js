import React, { useEffect, useLayoutEffect, useRef, useState } from 'react'
import { combineLatest, map, timer } from 'rxjs'
import sensorA from './sensors/sensorA'
import sensorB from './sensors/sensorB'
import sensorC from './sensors/sensorC'
import sensorD from './sensors/sensorD'

const Dashboard = () => {
    const valueRef = useRef();
    const [viewModel, setViewModel] = useState([])

    useLayoutEffect(() => {
        const observables = [sensorA.create(), sensorB.create(), sensorC.create(), sensorD.create()]
        const combined = combineLatest(observables)
        let subscribe = combined.pipe(
            map(vArray =>
                // ‘no data’ filter
                vArray.filter(v => {
                    let timeDiff = new Date().getTime() - v.lastRunTime
                    if (timeDiff > 1300) v.data = 'no data'
                    return true
                })
            ),
            map(v => valueRef.current = v)
        ).subscribe()
        return () => subscribe.unsubscribe()
    }, [])

    useEffect(() => {
        // view model refresh min 200 ms
        let subscribeId = timer(0, 200).pipe(map(i => {
            setViewModel(valueRef.current)
        })).subscribe()

        return () => subscribeId.unsubscribe()
    }, [])

    return (
        <div>
            <b>Welcome to the Dashboard!</b>
            {viewModel && viewModel.map(sensor =>
                <p key={sensor.name}> {sensor.name}: {sensor.data} </p>
            )}
        </div>
    )
}

export default Dashboard