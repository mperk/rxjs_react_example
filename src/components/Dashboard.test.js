import { cleanup } from "@testing-library/react";
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import { combineLatest, delay, first, firstValueFrom, lastValueFrom, merge, Observable, of, timer } from "rxjs";
import Dashboard from './Dashboard'
import sensorA from "./sensors/sensorA";
import sensorB from "./sensors/sensorB";
import sensorC from "./sensors/sensorC";
import sensorD from "./sensors/sensorD";

describe("Dashboard testing", () => {
    const observables = [sensorA.create(), sensorB.create(), sensorC.create(), sensorD.create()]
    const combined = combineLatest(observables)
    it('timer is working as expected', async () => {
        await expect(firstValueFrom(combined))
            .resolves.toBeTruthy();

        // All 4 sensors must emit at least one value
        await expect(firstValueFrom(combined))
            .resolves.toHaveLength(4)
    })

    it('timer and sensor testing after 1501 miliseconds', () => {
        jest.useFakeTimers()
        const nextFn = jest.fn()

        combined.subscribe({
            next: value => nextFn(value)
        })

        jest.advanceTimersByTime(1501);

        expect(nextFn).toHaveBeenCalledWith(
            expect.arrayContaining([
                {
                    name: expect.stringContaining("Sensor A"),
                    lastRunTime: expect.any(Number),
                    data: expect.any(String)
                },
                {
                    name: expect.stringContaining("Sensor B"),
                    lastRunTime: expect.any(Number),
                    data: expect.any(String)
                },
                {
                    name: expect.stringContaining("Sensor C"),
                    lastRunTime: expect.any(Number),
                    data: expect.any(String)
                },
                {
                    name: expect.stringContaining("Sensor D"),
                    lastRunTime: expect.any(Number),
                    data: expect.any(String)
                }])
        )
    })

    afterEach(cleanup);

    it('component testing', done => {
        jest.useFakeTimers()
        const wrapper = mount(<Dashboard />)
        let value = wrapper.find("p")
        expect(value).not.toBeUndefined()

        act(() => jest.advanceTimersByTime(1501))

        wrapper.update()

        const html = wrapper.find("div").text()
        expect(html).toContain("Sensor A")
        expect(html).toContain("Sensor B")
        expect(html).toContain("Sensor C")
        expect(html).toContain("Sensor D")

        expect(html).toEqual(wrapper.find("div").text())

        act(() => jest.advanceTimersByTime(1501))
        wrapper.update()
        
        const html2 = wrapper.find("div").text()
        expect(html).not.toEqual(html2)
        done()
    })

})


