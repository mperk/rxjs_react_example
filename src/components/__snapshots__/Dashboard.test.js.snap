// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component testing 1`] = `
<div>
  <b>
    Welcome to the Dashboard!
  </b>
  <p>
    Sensor A : 
  </p>
  <p>
    Sensor B : 
  </p>
  <p>
    Sensor C : 
  </p>
  <p>
    Sensor D : 
  </p>
</div>
`;
